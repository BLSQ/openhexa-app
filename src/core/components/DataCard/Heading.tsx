import clsx from "clsx";
import {
  getValue,
  ItemInstance,
  useItemContext,
  ValueAccessor,
} from "core/hooks/useItemContext";
import { ReactElement, useMemo } from "react";
import Block from "../Block";

type HeadingProps<T> = {
  titleAccessor?: ValueAccessor;
  subtitleAccessor?: ValueAccessor;
  className?: string;
  renderActions?(item: T): ReactElement;
  children?(item: T): ReactElement;
};

function Heading<T extends ItemInstance>(props: HeadingProps<T>) {
  const {
    titleAccessor,
    subtitleAccessor,
    renderActions,
    className,
    children,
  } = props;
  const { item } = useItemContext<T>();

  const title = useMemo(
    () => getValue(item, titleAccessor),
    [item, titleAccessor]
  );
  const subtitle = useMemo(
    () => (subtitleAccessor ? getValue(item, subtitleAccessor) : null),
    [item, subtitleAccessor]
  );

  return (
    <Block.Header
      className={clsx(className, "flex items-center justify-between")}
    >
      {children ? (
        children(item)
      ) : (
        <div>
          {title ?? ""}
          {subtitle && (
            <div className="truncate pt-2 text-sm text-gray-500">
              {subtitle}
            </div>
          )}
        </div>
      )}
      {renderActions ? renderActions(item) : null}
    </Block.Header>
  );
}

export default Heading;
