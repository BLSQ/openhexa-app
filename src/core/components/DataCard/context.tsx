import * as React from "react";
import { DataCardSectionInstance, Property, PropertyDefinition } from "./types";

export const DataCardSectionContext =
  React.createContext<DataCardSectionInstance | null>(null);

export function useDataCardSection() {
  const section = React.useContext(DataCardSectionContext);
  if (!section) {
    throw new Error(
      "useDataCardSection must be wrapped by a DataCard.Section."
    );
  }

  return section;
}

export function useDataCardProperty<V = any, FV = V>(
  definition: PropertyDefinition
): { property: Property<V, FV>; section: DataCardSectionInstance } {
  const section = useDataCardSection();
  const property = section.setProperty(definition);

  return { property, section };
}
