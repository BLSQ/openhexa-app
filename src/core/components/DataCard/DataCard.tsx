import clsx from "clsx";
import { ItemProvider } from "core/hooks/useItemContext";
import { ReactNode } from "react";
import Block from "../Block";
import Heading from "./Heading";
import PropertyDisplay from "./PropertyDisplay";
import Section from "./Section";

export type DataCardProps = {
  item: any;
  children: ReactNode;
  className?: string;
};

const DataCard = (props: DataCardProps) => {
  const { children, className, item } = props;
  return (
    <ItemProvider item={item}>
      <Block className={clsx("divide-y divide-gray-200", className)}>
        {children}
      </Block>
    </ItemProvider>
  );
};

DataCard.Section = Section;
DataCard.Property = PropertyDisplay;
DataCard.Heading = Heading;

export default DataCard;
