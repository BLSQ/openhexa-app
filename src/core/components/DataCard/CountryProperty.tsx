import CountryBadge from "core/features/CountryBadge";
import CountryPicker from "core/features/CountryPicker";
import { ensureArray } from "core/helpers/array";
import { useMemo } from "react";
import { useDataCardProperty } from "./context";
import DataCard from "./DataCard";
import { PropertyDefinition } from "./types";

type CountryPropertyProps = { multiple?: boolean } & PropertyDefinition;

const CountryProperty = (props: CountryPropertyProps) => {
  const { multiple, defaultValue, ...delegated } = props;
  const { property, section } = useDataCardProperty(delegated);

  const countriesArray = useMemo(
    () => ensureArray(property.displayValue),
    [property]
  );

  return (
    <DataCard.Property property={property}>
      {section.isEdited ? (
        <CountryPicker
          value={property.formValue ?? null}
          onChange={(v) => property.setValue(v)}
          multiple={multiple}
        />
      ) : (
        <div className="flex flex-wrap items-center gap-1.5">
          {countriesArray.length === 0 && defaultValue}
          {countriesArray.map((country, i) => (
            <CountryBadge key={i} country={country} />
          ))}
        </div>
      )}
    </DataCard.Property>
  );
};

export default CountryProperty;
