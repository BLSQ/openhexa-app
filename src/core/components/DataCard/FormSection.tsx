import {
  ChevronDownIcon,
  ChevronRightIcon,
  PencilIcon,
} from "@heroicons/react/24/outline";

import { BlockSection } from "core/components/Block";
import useForm, { FormInstance } from "core/hooks/useForm";
import {
  getValue,
  ItemInstance,
  useItemContext,
} from "core/hooks/useItemContext";
import _ from "lodash";
import { useTranslation } from "next-i18next";
import {
  ReactElement,
  ReactNode,
  useCallback,
  useEffect,
  useRef,
  useState,
} from "react";
import Button from "../Button/Button";
import DescriptionList, { DescriptionListProps } from "../DescriptionList";
import DisableClickPropagation from "../DisableClickPropagation";
import Spinner from "../Spinner";
import { DataCardSectionContext } from "./context";
import { Property, PropertyDefinition, PropertyFlag } from "./types";

export type OnSaveFn = (
  values: { [key: string]: any },
  item: ItemInstance
) => Promise<void> | void;

type FormSectionProps = {
  editLabel?: string;
  editIcon?: ReactElement;
  onSave?: OnSaveFn;
  title?: string;
  children: ReactNode;
} & Pick<DescriptionListProps, "displayMode" | "columns"> &
  Omit<React.ComponentProps<typeof BlockSection>, "title" | "children">;

function getPropertyFlag<F>(
  flag: PropertyFlag | undefined,
  displayValue: any,
  isEdited: boolean = false,
  form: FormInstance<F>
) {
  if (typeof flag === "function") {
    return flag(displayValue, isEdited, form.formData);
  }
  return flag;
}

function getProperty<F>(
  definition: PropertyDefinition,
  item: ItemInstance,
  form: FormInstance<F>,
  isEdited: boolean
) {
  const displayValue = getValue(item, definition.accessor);
  const prop: Property = {
    displayValue,
    defaultValue: definition.defaultValue,
    formValue: form.formData[definition.id as keyof F],
    validate: definition.validate,
    setValue: (value: any) =>
      form.setFieldValue(definition.id as keyof F, value),
    id: definition.id,
    label: definition.label,
    help: definition.help,
    hideLabel: definition.hideLabel ?? false,
    readonly:
      getPropertyFlag<F>(definition.readonly, displayValue, isEdited, form) ??
      false,
    required:
      getPropertyFlag<F>(definition.required, displayValue, isEdited, form) ??
      false,
    visible:
      getPropertyFlag<F>(definition.visible, displayValue, isEdited, form) ??
      true,
  };
  return prop;
}

function FormSection<F extends { [key: string]: any }>(
  props: FormSectionProps
) {
  const { t } = useTranslation();
  const {
    title,
    editLabel,
    editIcon,
    className,
    displayMode,
    columns,
    collapsible,
    defaultOpen,
    children,
    onSave,
  } = props;
  const { item } = useItemContext();

  const [isEdited, setEdited] = useState<boolean>(false);
  const toggleEdit = useCallback(() => setEdited((isEdited) => !isEdited), []);

  const definitions = useRef<PropertyDefinition[]>([]);
  const properties = useRef<{ [key: Property["id"]]: Property }>({});

  const form = useForm<F>({
    getInitialState() {
      const initialState = {} as any;
      Object.entries(properties.current).forEach(([key, prop]) => {
        initialState[key] = prop.displayValue;
      });
      return initialState;
    },
    onSubmit: async (values) => {
      if (onSave) {
        await onSave(values, item);
      }
      setEdited(false);
    },
    validate(values) {
      const errors = {} as any;

      for (const property of Object.values(properties.current)) {
        if (property.required && !values[property.id]) {
          errors[property.id] = t("This field is required");
        }
        if (property.validate) {
          const error = property.validate(values[property.id], values);
          if (error) {
            errors[property.id] = error;
          }
        }
      }
      return errors;
    },
  });

  useEffect(() => {
    if (isEdited) {
      form.resetForm();
    }
  }, [form, isEdited]);

  useEffect(() => {
    properties.current = definitions.current.reduce<{
      [key: string]: Property;
    }>((acc, def) => {
      acc[def.id] = getProperty<F>(def, item, form, isEdited);
      return acc;
    }, {});
    // form.validate();
  }, [definitions, item, form, form.formData, isEdited]);

  const section = {
    item,
    isEdited,
    toggleEdit,
    properties: properties.current,
    form,
    setProperty(definition: PropertyDefinition) {
      const existingDefinition = definitions.current.find(
        (x) => x.id === definition.id
      );
      if (!_.isEqual(definition, existingDefinition)) {
        definitions.current = definitions.current
          .filter((d) => d.id !== definition.id)
          .concat(definition);
      }
      return getProperty(definition, item, form, isEdited);
    },
  };

  return (
    <DataCardSectionContext.Provider value={section}>
      <BlockSection
        collapsible={collapsible}
        defaultOpen={defaultOpen}
        className={className}
        title={({ open }) => (
          <>
            <h4 className="font-medium">{title}</h4>
            {onSave && open && !isEdited && (
              <DisableClickPropagation>
                <button
                  className="ml-4 inline-flex items-center gap-1 text-sm text-blue-500 hover:text-blue-400"
                  onClick={toggleEdit}
                >
                  {editLabel ?? t("Edit")}
                  {editIcon ?? <PencilIcon className="h-4" />}
                </button>
              </DisableClickPropagation>
            )}
            <div className="flex flex-1 flex-shrink items-center justify-end">
              {collapsible && (
                <button title={open ? t("Hide") : t("Show")}>
                  {open ? (
                    <ChevronDownIcon className="h-5 w-5" />
                  ) : (
                    <ChevronRightIcon className="h-5 w-5" />
                  )}
                </button>
              )}
            </div>
          </>
        )}
      >
        {() =>
          isEdited ? (
            <form onSubmit={form.handleSubmit}>
              <DescriptionList columns={columns} displayMode={displayMode}>
                {children}
              </DescriptionList>

              {form.submitError && (
                <p className={"my-2 text-sm text-red-600"}>
                  {form.submitError}
                </p>
              )}
              <div className="mt-6 flex items-center justify-end gap-2">
                <Button type="submit" disabled={form.isSubmitting}>
                  {form.isSubmitting && <Spinner size="xs" className="mr-1" />}
                  {t("Save")}
                </Button>
                <Button onClick={toggleEdit} variant="white">
                  {t("Cancel")}
                </Button>
              </div>
            </form>
          ) : (
            <DescriptionList columns={columns} displayMode={displayMode}>
              {children}
            </DescriptionList>
          )
        }
      </BlockSection>
    </DataCardSectionContext.Provider>
  );
}

export default FormSection;
