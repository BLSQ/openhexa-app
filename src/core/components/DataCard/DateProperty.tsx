import { DateTimeFormatOptions } from "luxon";
import { useMemo } from "react";
import Input from "../forms/Input";
import Time from "../Time";
import { useDataCardProperty } from "./context";
import DataCard from "./DataCard";
import { PropertyDefinition } from "./types";

type DatePropertyProps = PropertyDefinition & {
  format?: DateTimeFormatOptions;
  min?: string;
  max?: string;
  markdown?: boolean;
};

const DateProperty = (props: DatePropertyProps) => {
  const { format, min, max, ...delegated } = props;

  const { property, section } = useDataCardProperty(delegated);

  const value = useMemo(() => {
    return property.formValue?.split(".")[0];
  }, [property.formValue]);

  return (
    <DataCard.Property property={property}>
      {section.isEdited && !property.readonly ? (
        <Input
          name={property.id}
          type="datetime-local"
          required={property.required}
          value={value}
          min={min}
          max={max}
          onChange={(e) => property.setValue(e.target.value)}
        />
      ) : (
        property.displayValue && (
          <Time datetime={property.displayValue} format={format} />
        )
      )}
    </DataCard.Property>
  );
};

export default DateProperty;
