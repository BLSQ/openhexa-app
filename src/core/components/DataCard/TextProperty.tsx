import ReactMarkdown from "react-markdown";
import Input from "../forms/Input";
import Textarea from "../forms/Textarea";
import { useDataCardProperty } from "./context";
import DataCard from "./DataCard";
import { PropertyDefinition } from "./types";

type TextPropertyProps = PropertyDefinition & {
  markdown?: boolean;
  defaultValue?: string;
};

const TextProperty = (props: TextPropertyProps) => {
  const { markdown, defaultValue, id, accessor, required, readonly, label } =
    props;

  const { property, section } = useDataCardProperty({
    id,
    accessor,
    required,
    readonly,
    label,
  });

  if (section.isEdited && !property.readonly) {
    return (
      <DataCard.Property property={property}>
        {markdown ? (
          <Textarea
            className="w-full"
            value={property.formValue}
            onChange={(e) => property.setValue(e.target.value)}
            required={property.required}
            readOnly={property.readonly}
          />
        ) : (
          <Input
            value={property.formValue ?? ""}
            onChange={(e) => property.setValue(e.target.value)}
            required={property.required}
            readOnly={property.readonly}
          />
        )}
      </DataCard.Property>
    );
  } else {
    return (
      <DataCard.Property property={property}>
        {markdown && property.displayValue ? (
          <ReactMarkdown className="prose max-w-3xl text-sm">
            {property.displayValue}
          </ReactMarkdown>
        ) : (
          <div className="prose text-sm">
            {property.displayValue || defaultValue}
          </div>
        )}
      </DataCard.Property>
    );
  }
};

export default TextProperty;
