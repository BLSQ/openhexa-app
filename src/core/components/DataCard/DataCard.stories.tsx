import { Story } from "@ladle/react";
import BadgeProperty from "./BadgeProperty";

import DataCard, { DataCardProps } from "./DataCard";
import Section from "./Section";
import TextProperty from "./TextProperty";

interface DataCardStoryProps extends DataCardProps {}

export const DataCardStory: Story<DataCardStoryProps> = ({ children }) => {
  return (
    <DataCard
      item={{
        name: "Alfonse",
        description: `### Description Field

Hey It's in *italic*!`,
        tags: [{ label: "Hey" }, { label: "Tag" }],
      }}
    >
      <DataCard.Heading titleAccessor="name" />
      <DataCard.Section title="First Section">
        <TextProperty id="name" label="Name" accessor="name" />
        <TextProperty
          id="description"
          label="Description"
          accessor="description"
          markdown
        />
        <BadgeProperty id="tags" label="Badges" accessor="tags" />
      </DataCard.Section>
    </DataCard>
  );
};

DataCardStory.storyName = "DataCard";

const defaults = {
  title: "UI / DataCard",
};
export default defaults;
