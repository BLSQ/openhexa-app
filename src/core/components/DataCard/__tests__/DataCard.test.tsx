import DataCard from "../DataCard";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { faker } from "@faker-js/faker";
import TextProperty from "../TextProperty";

function nextFrame() {
  return new Promise<void>((resolve) => {
    requestAnimationFrame(() => {
      requestAnimationFrame(() => {
        resolve();
      });
    });
  });
}

describe("DataCard", () => {
  it("renders", async () => {
    const item = {
      name: faker.name.jobDescriptor(),
    };
    const { container } = render(
      <DataCard item={item}>
        <DataCard.Heading titleAccessor="name" />
        <DataCard.FormSection>
          <TextProperty id="name" label="Name" accessor="name" />
        </DataCard.FormSection>
      </DataCard>
    );

    expect(container).toMatchSnapshot();
  });
});
