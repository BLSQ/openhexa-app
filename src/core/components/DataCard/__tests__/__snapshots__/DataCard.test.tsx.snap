// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataCard renders 1`] = `
<div>
  <article
    class="sm:rounded-lg overflow-hidden shadow border-b border-gray-200 bg-white divide-y divide-gray-200"
  >
    <h3
      class="px-4 py-5 pt-7 text-lg font-medium leading-4 text-gray-900 sm:px-6 flex items-center justify-between"
    >
      <div>
        Regional
      </div>
    </h3>
    <section
      class="px-4 py-5 sm:px-6"
    >
      <div>
        <div
          aria-controls="headlessui-disclosure-panel-67862f3c-cbfc-451b-98ed-ed1d0420ea19"
          aria-expanded="true"
          class="-my-7 flex items-center py-7"
          data-headlessui-state="open"
          id="headlessui-disclosure-button-fbe024f2-3162-465a-b6e8-d65a837e308a"
        >
          <div
            class="flex flex-1 flex-shrink items-center justify-end"
          />
        </div>
        <div>
          <div
            class="mt-6"
            data-headlessui-state="open"
            id="headlessui-disclosure-panel-67862f3c-cbfc-451b-98ed-ed1d0420ea19"
          >
            <dl
              class="grid gap-4"
            >
              <div
                class="grid grid-cols-5 gap-2 sm:gap-4"
              >
                <dt
                  class="flex text-sm font-medium text-gray-500 col-span-1 items-center"
                >
                  <span
                    class="inline"
                  >
                    Name
                  </span>
                </dt>
                <dd
                  class="col-span-4 text-sm text-gray-900"
                >
                  <div
                    class="prose text-sm text-gray-900"
                  >
                    Regional
                  </div>
                </dd>
              </div>
            </dl>
          </div>
        </div>
      </div>
    </section>
  </article>
</div>
`;
